File: frontend/src/utils/errorHandler.js
// Centralized error handling utilities - generic for all websites

export const ERROR_TYPES = {
  NETWORK: 'network',
  AUTH: 'authentication',
  VALIDATION: 'validation',
  SERVER: 'server',
  TIMEOUT: 'timeout',
  PARSE: 'parse'
};

// Parse backend error responses to user-friendly messages
export const parseBackendError = (error, response = null) => {
  // Network errors
  if (!navigator.onLine) {
    return {
      type: ERROR_TYPES.NETWORK,
      title: 'Connection Error',
      message: 'You appear to be offline. Please check your internet connection.',
      canRetry: true
    };
  }

  // Request timeout
  if (error?.name === 'AbortError' || error?.message?.includes('timeout')) {
    return {
      type: ERROR_TYPES.TIMEOUT,
      title: 'Request Timeout',
      message: 'The request took too long. Please try again.',
      canRetry: true
    };
  }

  // Network/fetch errors
  if (error?.name === 'TypeError' || error?.message?.includes('fetch')) {
    return {
      type: ERROR_TYPES.NETWORK,
      title: 'Network Error',
      message: 'Unable to connect to the server. Please check your connection and try again.',
      canRetry: true
    };
  }

  if (!response) {
    return {
      type: ERROR_TYPES.SERVER,
      title: 'Unexpected Error',
      message: error?.message || 'Something went wrong. Please try again.',
      canRetry: true
    };
  }

  // Parse response-based errors
  const status = response.status;
  
  if (status === 401) {
    return {
      type: ERROR_TYPES.AUTH,
      title: 'Authentication Required',
      message: 'Please log in to continue.',
      canRetry: false
    };
  }

  if (status === 403) {
    return {
      type: ERROR_TYPES.AUTH,
      title: 'Access Denied',
      message: 'You don\'t have permission to perform this action.',
      canRetry: false
    };
  }

  if (status === 404) {
    return {
      type: ERROR_TYPES.SERVER,
      title: 'Not Found',
      message: 'The requested resource was not found.',
      canRetry: false
    };
  }

  if (status === 429) {
    return {
      type: ERROR_TYPES.SERVER,
      title: 'Too Many Requests',
      message: 'Please slow down. Try again in a few moments.',
      canRetry: true
    };
  }

  if (status >= 500) {
    return {
      type: ERROR_TYPES.SERVER,
      title: 'Server Error',
      message: 'The server is experiencing issues. Please try again later.',
      canRetry: true
    };
  }

  if (status >= 400) {
    return {
      type: ERROR_TYPES.VALIDATION,
      title: 'Request Error',
      message: 'There was an issue with your request. Please check your input and try again.',
      canRetry: false
    };
  }

  return {
    type: ERROR_TYPES.SERVER,
    title: 'Unknown Error',
    message: 'An unexpected error occurred. Please try again.',
    canRetry: true
  };
};

// Extract error message from backend response
export const extractErrorMessage = async (response) => {
  try {
    const data = await response.json();
    
    // New standardized backend error format (utils/error_handler.py)
    if (data.error && typeof data.error === 'object') {
      const error = data.error;
      let message = error.message || 'An error occurred';
      
      // Add details if available
      if (error.details) {
        if (typeof error.details === 'string') {
          message += ` (${error.details})`;
        } else if (error.details.validation_errors) {
          // Parse validation errors
          const validationErrors = parseValidationErrors(error.details.validation_errors);
          message = validationErrors;
        } else if (Object.keys(error.details).length > 0) {
          // Show first detail
          const firstDetail = Object.values(error.details)[0];
          if (typeof firstDetail === 'string') {
            message += ` (${firstDetail})`;
          }
        }
      }
      
      return message;
    }
    
    // Django REST Framework error formats
    if (data.error) return data.error;
    if (data.detail) return data.detail;
    if (data.message) return data.message;
    
    // Validation errors
    if (data.non_field_errors) return data.non_field_errors[0];
    
    // Field-specific errors (DRF serializer errors)
    const fieldErrors = Object.keys(data).filter(key => 
      Array.isArray(data[key]) && data[key].length > 0
    );
    if (fieldErrors.length > 0) {
      const field = fieldErrors[0];
      const fieldName = formatFieldName(field);
      return `${fieldName}: ${data[field][0]}`;
    }

    return 'An error occurred';
  } catch (e) {
    return 'Unable to parse error response';
  }
};

// Parse validation errors into readable format
const parseValidationErrors = (errors) => {
  if (typeof errors === 'string') return errors;
  
  if (Array.isArray(errors)) {
    return errors.join(', ');
  }
  
  if (typeof errors === 'object') {
    const messages = [];
    for (const [field, fieldErrors] of Object.entries(errors)) {
      const fieldName = formatFieldName(field);
      if (Array.isArray(fieldErrors)) {
        fieldErrors.forEach(err => {
          if (field === 'non_field_errors') {
            messages.push(err);
          } else {
            messages.push(`${fieldName}: ${err}`);
          }
        });
      } else {
        messages.push(`${fieldName}: ${fieldErrors}`);
      }
    }
    return messages.join('; ') || 'Validation error occurred';
  }
  
  return 'Validation error occurred';
};

// Format field names for display (snake_case to Title Case)
const formatFieldName = (field) => {
  if (field === 'non_field_errors') return 'Error';
  return field
    .split('_')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
};

// Create error object from response
export const createErrorFromResponse = async (error, response) => {
  const baseError = parseBackendError(error, response);
  
  if (response) {
    try {
      const message = await extractErrorMessage(response);
      return { ...baseError, message };
    } catch (e) {
      // If message extraction fails, return base error
      return baseError;
    }
  }
  
  return baseError;
};

// Format error for display to user
export const formatErrorForDisplay = (error) => {
  if (typeof error === 'string') return error;
  if (error?.message) return error.message;
  if (error?.title) return error.title;
  return 'An error occurred';
};

// Log error to console (can be extended to send to logging service)
export const logError = (error, context = '') => {
  const errorMessage = context ? `[${context}] ${error}` : error;
  console.error(errorMessage, error);
  
  // TODO: Send to logging service (Sentry, etc.)
  // if (window.Sentry) {
  //   window.Sentry.captureException(error, { context });
  // }
};

File: backend/utils/error_handler.py
"""
Centralized error handling utilities for consistent error responses.
Generic implementation for all websites.
"""
from __future__ import annotations

from typing import Any, Dict, Optional
from rest_framework.response import Response
from rest_framework import status
from django.db import IntegrityError, DatabaseError
from django.core.exceptions import ValidationError as DjangoValidationError
from rest_framework.exceptions import ValidationError as DRFValidationError
import logging

logger = logging.getLogger('utils')


class ErrorType:
    """Error type constants"""
    VALIDATION_ERROR = "validation_error"
    NOT_FOUND = "not_found"
    PERMISSION_DENIED = "permission_denied"
    AUTHENTICATION_ERROR = "authentication_error"
    DATABASE_ERROR = "database_error"
    INTEGRITY_ERROR = "integrity_error"
    SERVER_ERROR = "server_error"
    BAD_REQUEST = "bad_request"
    EXTERNAL_API_ERROR = "external_api_error"
    CONFLICT = "conflict"


def error_response(
    error_type: str,
    message: str,
    details: Optional[Dict[str, Any]] = None,
    status_code: int = status.HTTP_400_BAD_REQUEST,
    log_error: bool = True
) -> Response:
    """
    Create a standardized error response.
    
    Args:
        error_type: Type of error (use ErrorType constants)
        message: Human-readable error message
        details: Additional error details/context
        status_code: HTTP status code
        log_error: Whether to log the error
    
    Returns:
        Response object with error details
    """
    error_data = {
        "success": False,
        "error": {
            "type": error_type,
            "message": message,
        }
    }
    
    if details:
        error_data["error"]["details"] = details
    
    if log_error:
        log_message = f"{error_type}: {message}"
        if details:
            log_message += f" | Details: {details}"
        
        if status_code >= 500:
            logger.error(log_message)
        elif status_code >= 400:
            logger.warning(log_message)
        else:
            logger.info(log_message)
    
    return Response(error_data, status=status_code)


def success_response(
    message: str,
    data: Optional[Any] = None,
    status_code: int = status.HTTP_200_OK
) -> Response:
    """
    Create a standardized success response.
    
    Args:
        message: Success message
        data: Response data
        status_code: HTTP status code
    
    Returns:
        Response object with success data
    """
    response_data = {
        "success": True,
        "message": message,
    }
    
    if data is not None:
        response_data["data"] = data
    
    return Response(response_data, status=status_code)


def handle_exception(
    exception: Exception,
    context: str = "Operation",
    user_message: Optional[str] = None
) -> Response:
    """
    Handle common exceptions and return appropriate error response.
    
    Args:
        exception: The exception that occurred
        context: Context description for logging
        user_message: Custom user-facing message (optional)
    
    Returns:
        Response object with error details
    """
    logger.error(f"{context} failed: {type(exception).__name__}: {str(exception)}")
    
    # Database integrity errors (unique constraint, foreign key, etc.)
    if isinstance(exception, IntegrityError):
        error_msg = user_message or "A database constraint was violated. This item may already exist."
        details = {"db_error": str(exception)}
        
        # Check for specific integrity errors
        if "unique constraint" in str(exception).lower():
            error_msg = user_message or "An item with this value already exists."
            details["constraint"] = "unique"
        elif "foreign key" in str(exception).lower():
            error_msg = user_message or "Referenced item does not exist."
            details["constraint"] = "foreign_key"
        elif "not null" in str(exception).lower():
            error_msg = user_message or "Required field is missing."
            details["constraint"] = "not_null"
        
        return error_response(
            ErrorType.INTEGRITY_ERROR,
            error_msg,
            details=details,
            status_code=status.HTTP_409_CONFLICT
        )
    
    # General database errors
    if isinstance(exception, DatabaseError):
        return error_response(
            ErrorType.DATABASE_ERROR,
            user_message or "A database error occurred. Please try again.",
            details={"db_error": str(exception)},
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )
    
    # Django validation errors
    if isinstance(exception, DjangoValidationError):
        return error_response(
            ErrorType.VALIDATION_ERROR,
            user_message or "Validation error occurred.",
            details={"validation_errors": exception.message_dict if hasattr(exception, 'message_dict') else str(exception)},
            status_code=status.HTTP_400_BAD_REQUEST
        )
    
    # DRF validation errors
    if isinstance(exception, DRFValidationError):
        return error_response(
            ErrorType.VALIDATION_ERROR,
            user_message or "Invalid data provided.",
            details={"validation_errors": exception.detail},
            status_code=status.HTTP_400_BAD_REQUEST
        )
    
    # Value errors (often from business logic)
    if isinstance(exception, ValueError):
        return error_response(
            ErrorType.BAD_REQUEST,
            user_message or str(exception),
            status_code=status.HTTP_400_BAD_REQUEST
        )
    
    # Permission errors
    if isinstance(exception, PermissionError):
        return error_response(
            ErrorType.PERMISSION_DENIED,
            user_message or "You don't have permission to perform this action.",
            status_code=status.HTTP_403_FORBIDDEN
        )
    
    # Generic exception - server error
    return error_response(
        ErrorType.SERVER_ERROR,
        user_message or "An unexpected error occurred. Please try again later.",
        details={"error": str(exception), "type": type(exception).__name__},
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
    )


def parse_serializer_errors(serializer_errors: dict) -> str:
    """
    Parse DRF serializer errors into a readable message.
    
    Args:
        serializer_errors: Serializer error dict
    
    Returns:
        Human-readable error message
    """
    messages = []
    
    for field, errors in serializer_errors.items():
        if isinstance(errors, list):
            for error in errors:
                if field == 'non_field_errors':
                    messages.append(str(error))
                else:
                    messages.append(f"{field}: {error}")
        else:
            messages.append(f"{field}: {errors}")
    
    return "; ".join(messages) if messages else "Validation failed"


def ratelimit_exception_handler(exc, context):
    """
    Custom exception handler for django-ratelimit.
    Catches Ratelimited exceptions and returns proper 429 response.
    """
    from django_ratelimit.exceptions import Ratelimited
    from rest_framework.views import exception_handler
    
    # Call DRF's default exception handler first
    response = exception_handler(exc, context)
    
    # Check if this is a rate limit exception
    if isinstance(exc, Ratelimited):
        logger.warning(f"Rate limit exceeded for IP: {context['request'].META.get('REMOTE_ADDR')} on {context['request'].path}")
        
        return error_response(
            ErrorType.BAD_REQUEST,
            "Too many requests. Please slow down and try again later.",
            details={
                "rate_limit_exceeded": True,
                "retry_after": "Please wait a few moments before trying again."
            },
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            log_error=False  # Already logged above
        )
    
    return response


def ratelimit_error_response(request, exception=None):
    """
    Custom response for rate limit exceeded errors.
    Called by django-ratelimit when rate limit is exceeded.
    
    Args:
        request: Django request object
        exception: Optional exception object
    
    Returns:
        Response object with rate limit error
    """
    logger.warning(f"Rate limit exceeded for IP: {request.META.get('REMOTE_ADDR')} on {request.path}")
    
    return error_response(
        ErrorType.BAD_REQUEST,
        "Too many requests. Please slow down and try again later.",
        details={
            "rate_limit_exceeded": True,
            "retry_after": "Please wait a few moments before trying again."
        },
        status_code=status.HTTP_429_TOO_MANY_REQUESTS,
        log_error=False  # Already logged above
    )
